freqtrade/freqtradebot.py

```diff
@@ -1140,33 +1141,85 @@ def check_handle_timedout(self) -> None:
+        if order['side'] == trade.entry_side:
+            self.handle_cancel_enter(trade, order, constants.CANCEL_REASON['TIMEOUT'])
+        else:
+            canceled = self.handle_cancel_exit(
+                trade, order, constants.CANCEL_REASON['TIMEOUT'])
+            canceled_count = trade.get_exit_order_count()
+            max_timeouts = self.config.get('unfilledtimeout', {}).get('exit_timeout_count', 0)
+            if canceled and max_timeouts > 0 and canceled_count >= max_timeouts:
+                logger.warning(f'Emergency exiting trade {trade}, as the exit order '
+                               f'timed out {max_timeouts} times.')
+                try:
+                    self.execute_trade_exit(
+                        trade, order['price'],
+                        exit_check=ExitCheckTuple(exit_type=ExitType.EMERGENCY_EXIT))
+                except DependencyException as exception:
+                    logger.warning(
+                        f'Unable to emergency sell trade {trade.pair}: {exception}')
+
+    def replace_order(self, order: Dict, order_obj: Optional[Order], trade: Trade) -> None:
+        """
+        Check if current analyzed entry order should be replaced. Analyzed order is canceled
+        if adjust_entry_price() returned price differs from proposed_rate.
+        New order is only placed if adjust_entry_price() returned price is not None.
+        :param order: Order dict grabbed with exchange.fetch_order()
+        :param order_obj: Order object.
+        :param trade: Trade object.
+        :return: None
+        """
+        analyzed_df, _ = self.dataprovider.get_analyzed_dataframe(trade.pair,
+                                                                  self.strategy.timeframe)
+        latest_candle_open_date = analyzed_df.iloc[-1]['date'] if len(analyzed_df) > 0 else None
+        latest_candle_close_date = timeframe_to_next_date(self.strategy.timeframe,
+                                                          latest_candle_open_date)
+        # Check if new candle
+        if order_obj and latest_candle_close_date > order_obj.order_date_utc:
+            # New candle
+            proposed_rate = self.exchange.get_rate(
+                trade.pair, side='entry', is_short=trade.is_short, refresh=True)
+            adjusted_entry_price = strategy_safe_wrapper(self.strategy.adjust_entry_price,
+                                                         default_retval=proposed_rate)(
+                trade=trade, order=order_obj, pair=trade.pair,
+                current_time=datetime.now(timezone.utc), proposed_rate=proposed_rate,
+                entry_tag=trade.enter_tag, side=trade.entry_side)
+
+            full_cancel = False
+            cancel_reason = constants.CANCEL_REASON['REPLACE']
+            if not adjusted_entry_price:
+                full_cancel = True
+                cancel_reason = constants.CANCEL_REASON['USER_CANCEL']
+            if proposed_rate != adjusted_entry_price:
```

Comment:
**On current implementation:**
> 1. leave order as is
 Indeed user would have to return `proposed_rate`

> 2. cancel order from this callback (i assume that's None?)
Yup. User would have to return `None` to just cancel the order (with no replacement)

> 3. set new order at proposed rate
Can't do. As you've noticed it would need a slight modification to `proposed_rate`. Used to think that it's not a "big deal" but if I consider BTC pairs it might be a significant difference.

> 4. set the new order at any random given rate - say 225$
User simply returns the desired rate for the new order. If I'm missing something here regarding fees or such, I'm sorry, I'm probably oblivious to such things at the moment.

**Proposed solutions:**
1. Simple hackish (probably confusing a bit for end user?):
To leave order as is user would need to `return -proposed_stake`
To use the "new best rate" user would need to `return proposed_stake`

2. Add an extra `bool` return to better specify intent:
> 1. leave order as is
`return proposed_stake, False`
> 2. cancel order from this callback (i assume that's None?)
`return proposed_stake, None`
> 3. set new order at proposed rate
`return proposed_stake, True`
> 4. set the new order at any random given rate - say 225$
`return 225, True`

If there's a simpler/better way that I've missed while replying to this please let me know. As always thank you for the feedback and looking forward for your thoughts on my proposals.

Reply:
> User simply returns the desired rate for the new order. If I'm missing something here regarding fees or such, I'm sorry, I'm probably oblivious to such things at the moment.

fees don't matter here - we're talking about the order price - not profits.
my point in this was - if that new desired rate is (accidentally) the new proposed rate, it would also not work.

---

i think both solutions CAN confuse the user, the first due to random negative return values - the 2nd by using tuples (tuples in python are NOT that easy to grasp - and sometimes error-prone if you're not careful).

i think the big problem comes from supporting cancellation out of this callback. 
if you take this out - then `None` leaves the order in place - and every other return value sets a new rate.

what's the main thought behind supporting cancellation from here as well? it'll complicate both freqtrade's code (you now have 2 callbacks which can do the same thing) - and potentially confuse users (which one should i use??).

what benefit would it provide over the `custom_entry_timeout()` callback?

Reply:
> what benefit would it provide over the `custom_entry_timeout()` callback?

Nothing appart from slight convenience. I imagine whomever wants to use this with multiple branches and different timeouts for each would have some repeated logic both in `custom_entry_timeout()` and `adjust_entry_price()`. No that big of a deal probably.

But yeah, I see your point. Better to have redundant code in strat (with low probability of occuring anyway) than have it in freqtrade code.

Reply:
no one said defining such a thing is easy :laughing: 

the only thing i see as REALLY problematic at this point is to leave the order in place with `proposed_rate`. we can leave the order in place if the order price is returned (maybe this is the point we're looking for?)

Thinking about this a bit more - the current code would anyway replace the order in the wrong instance.
assume an open order at 99$ (it's not filled).
price now moved to 100$ (so the proposed rate is 100$)
if the strategy now return 99$ again - we'd currently cancel and replace the order - which is clearly unnecessary.
if instead, i return 100$ (to move the order higher) - then we'd NOT replace the order - which seems wrong (i'd have to send 100.0001$ to have it replaced).

so maybe the correct approach would be to pass in a "current_order_rate" - which acts as "leave order in place". Then i think we would be able to cover all of the above cases.

Reply:
Oh this sounds great!

Reply:
Addressed this through https://github.com/freqtrade/freqtrade/pull/6692/commits/17650d7e6025385f8d9dbd3dbe44d097b50ce9eb.